File: download.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
)

// ── Parallel segment download ─────────────────────────────────────────────

func downloadSegments(segments []Segment, tmpDir string, concurrency int, onProgress func(done, total int)) ([]string, error) {
	total := len(segments)
	segFiles := make([]string, total)
	var done int64
	var firstErr error
	var errOnce sync.Once

	sem := make(chan struct{}, concurrency)
	var wg sync.WaitGroup

	for i, seg := range segments {
		dest := filepath.Join(tmpDir, fmt.Sprintf("seg%06d.ts", i))
		segFiles[i] = dest

		wg.Add(1)
		sem <- struct{}{}
		go func(idx int, url, dest string) {
			defer wg.Done()
			defer func() { <-sem }()

			if err := curlDownload(url, dest); err != nil {
				errOnce.Do(func() { firstErr = fmt.Errorf("segment %d: %w", idx, err) })
				return
			}
			d := int(atomic.AddInt64(&done, 1))
			if onProgress != nil {
				onProgress(d, total)
			}
		}(i, seg.URL, dest)
	}
	wg.Wait()
	return segFiles, firstErr
}

// ── ffmpeg remux ──────────────────────────────────────────────────────────

func remux(segFiles []string, outputFile string, trimStart, totalDuration float64) error {
	tmpDir := filepath.Dir(segFiles[0])
	concatList := filepath.Join(tmpDir, "concat.txt")

	var lines []string
	for _, f := range segFiles {
		lines = append(lines, fmt.Sprintf("file '%s'", f))
	}
	if err := os.WriteFile(concatList, []byte(strings.Join(lines, "\n")), 0644); err != nil {
		return err
	}

	ffmpeg := findFFmpeg()
	args := []string{"-f", "concat", "-safe", "0", "-i", concatList}
	if trimStart > 0 {
		args = append(args, "-ss", fmt.Sprintf("%.3f", trimStart))
	}
	if totalDuration > 0 {
		args = append(args, "-t", fmt.Sprintf("%.3f", totalDuration))
	}
	args = append(args, "-c", "copy", "-movflags", "+faststart", "-y", outputFile)

	cmd := exec.Command(ffmpeg, args...)
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// ── Full VOD download pipeline ────────────────────────────────────────────

type DownloadResult struct {
	Size      int64
	Duration  float64
	Qualities []Stream
}

func downloadVOD(m3u8URL, quality, outputFile string, timeRange *TimeRange, onProgress func(done, total int), onStatus func(string)) (*DownloadResult, error) {
	onStatus("Fetching stream info...")
	masterText, err := fetchM3U8(m3u8URL)
	if err != nil {
		return nil, err
	}

	segPlaylistURL := m3u8URL
	var qualities []Stream

	if strings.Contains(masterText, "#EXT-X-STREAM-INF") {
		qualities = parseMasterPlaylist(masterText, m3u8URL)
		// Pick requested quality or best
		chosen := qualities[0].URL
		for _, s := range qualities {
			if strings.Contains(s.URL, quality) {
				chosen = s.URL
				onStatus(fmt.Sprintf("Quality: %s", s.Resolution))
				break
			}
		}
		if chosen == qualities[0].URL {
			onStatus(fmt.Sprintf("Quality: %s (best)", qualities[0].Resolution))
		}
		segPlaylistURL = chosen
	}

	onStatus("Fetching segment list...")
	segText, err := fetchM3U8(segPlaylistURL)
	if err != nil {
		return nil, err
	}
	segments := parseSegmentPlaylist(segText, segPlaylistURL)
	fullDur := totalDuration(segments)
	onStatus(fmt.Sprintf("Full duration: %s (%d segments)", formatDuration(fullDur), len(segments)))

	// Apply time range
	sel := selectSegments(segments, timeRange)
	if timeRange != nil {
		onStatus(fmt.Sprintf("Time range: %s → %s (%d segments)",
			formatTimestamp(timeRange.Start), formatTimestamp(timeRange.End), len(sel.Segments)))
	}

	// Download segments
	tmpDir, err := os.MkdirTemp("", "kick-")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(tmpDir)

	onStatus("Downloading segments...")
	segFiles, err := downloadSegments(sel.Segments, tmpDir, 8, onProgress)
	if err != nil {
		return nil, err
	}

	onStatus("Remuxing to MP4...")
	if err := remux(segFiles, outputFile, sel.TrimStart, sel.TotalDuration); err != nil {
		return nil, fmt.Errorf("ffmpeg: %w", err)
	}

	fi, _ := os.Stat(outputFile)
	size := int64(0)
	if fi != nil {
		size = fi.Size()
	}

	dur := sel.TotalDuration
	if dur == 0 {
		dur = fullDur
	}
	return &DownloadResult{Size: size, Duration: dur, Qualities: qualities}, nil
}

// ── Live stream recording ─────────────────────────────────────────────────

func recordLive(m3u8URL, quality, outputFile string, duration float64, onStatus func(string)) error {
	// Pick quality from master
	masterText, err := fetchM3U8(m3u8URL)
	if err == nil && strings.Contains(masterText, "#EXT-X-STREAM-INF") {
		streams := parseMasterPlaylist(masterText, m3u8URL)
		for _, s := range streams {
			if strings.Contains(s.URL, quality) {
				m3u8URL = s.URL
				onStatus(fmt.Sprintf("Quality: %s", s.Resolution))
				break
			}
		}
	}

	ffmpeg := findFFmpeg()
	args := []string{
		"-user_agent", userAgent,
		"-headers", "Referer: https://kick.com/\r\n",
		"-rw_timeout", "10000000",
		"-i", m3u8URL,
	}
	if duration > 0 {
		args = append(args, "-t", fmt.Sprintf("%.0f", duration))
		onStatus(fmt.Sprintf("Recording %s → %s", formatDuration(duration), outputFile))
	} else {
		onStatus(fmt.Sprintf("Recording → %s (Ctrl+C to stop)", outputFile))
	}
	args = append(args, "-c", "copy", "-movflags", "+faststart", "-y", outputFile)

	cmd := exec.Command(ffmpeg, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// ── Clip download ─────────────────────────────────────────────────────────

func downloadClip(clipURL, outputFile string, onStatus func(string)) error {
	if strings.Contains(clipURL, ".mp4") {
		onStatus("Downloading clip...")
		return curlDownload(clipURL, outputFile)
	}
	// HLS clip
	_, err := downloadVOD(clipURL, "720p60", outputFile, nil,
		func(d, t int) { fmt.Printf("\r  seg %d/%d", d, t) },
		onStatus)
	fmt.Println()
	return err
}

File: src/download.js
/**
 * download.js — HLS parsing, segment downloading, time-range clipping, live recording
 */
const { spawnSync, spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { curlGet } = require('./api');
const { getFFmpegPath } = require('./setup');

const CURL_HEADERS = [
  '-H', 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
  '-H', 'Referer: https://kick.com/',
  '-H', 'Origin: https://kick.com',
];

// ── Time parsing ──────────────────────────────────────────────────────────
function parseTimestamp(str) {
  if (!str || !str.trim()) return null;
  const parts = str.trim().split(':').map(Number);
  if (parts.some(isNaN)) return null;
  if (parts.length === 3) return parts[0] * 3600 + parts[1] * 60 + parts[2];
  if (parts.length === 2) return parts[0] * 60 + parts[1];
  return parts[0];
}

function parseTimeRange(str) {
  if (!str || !str.trim()) return null;
  const [startStr, endStr] = str.split('-').map(s => s.trim());
  const start = parseTimestamp(startStr);
  const end = parseTimestamp(endStr);
  if (start === null || end === null || end <= start) return null;
  return { start, end };
}

function formatDuration(sec) {
  if (!sec || sec <= 0) return '0s';
  const h = Math.floor(sec / 3600);
  const m = Math.floor((sec % 3600) / 60);
  const s = Math.floor(sec % 60);
  if (h > 0) return `${h}h ${m}m ${s}s`;
  if (m > 0) return `${m}m ${s}s`;
  return `${s}s`;
}

function formatTimestamp(sec) {
  const h = Math.floor(sec / 3600);
  const m = Math.floor((sec % 3600) / 60);
  const s = Math.floor(sec % 60);
  return `${h}:${String(m).padStart(2,'0')}:${String(s).padStart(2,'0')}`;
}

// ── HLS parsing ───────────────────────────────────────────────────────────
function parseMasterPlaylist(text, baseUrl) {
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
  const streams = [];
  for (let i = 0; i < lines.length; i++) {
    if (!lines[i].startsWith('#EXT-X-STREAM-INF')) continue;
    const playlistLine = lines[i + 1];
    if (!playlistLine || playlistLine.startsWith('#')) continue;
    const bw  = lines[i].match(/BANDWIDTH=(\d+)/);
    const res = lines[i].match(/RESOLUTION=(\d+x\d+)/);
    const fps = lines[i].match(/FRAME-RATE=([\d.]+)/);
    const url = playlistLine.startsWith('http') ? playlistLine : new URL(playlistLine, baseUrl).href;
    // Derive quality label from URL path (e.g. "720p60")
    const labelMatch = playlistLine.match(/([\dp]+)/);
    streams.push({
      bandwidth:  bw ? parseInt(bw[1]) : 0,
      resolution: res ? res[1] : '?',
      fps:        fps ? parseFloat(fps[1]) : 30,
      url,
      label:      labelMatch ? labelMatch[1] : (res ? res[1] : '?'),
    });
  }
  streams.sort((a, b) => b.bandwidth - a.bandwidth);
  return streams;
}

function parseSegmentPlaylist(text, baseUrl) {
  const lines = text.split('\n').map(l => l.trim());
  const segments = [];
  let duration = 0;
  for (let i = 0; i < lines.length; i++) {
    const durMatch = lines[i].match(/^#EXTINF:([\d.]+)/);
    if (durMatch) {
      duration = parseFloat(durMatch[1]);
      continue;
    }
    if (lines[i] && !lines[i].startsWith('#')) {
      const url = lines[i].startsWith('http') ? lines[i] : new URL(lines[i], baseUrl).href;
      segments.push({ url, duration });
    }
  }
  return segments;
}

// ── Fetch & parse m3u8 ────────────────────────────────────────────────────
function fetchM3U8(url) {
  return curlGet(url, '*/*');
}

function getQualities(m3u8Url) {
  const text = fetchM3U8(m3u8Url);
  if (!text.includes('#EXTM3U')) throw new Error('Invalid m3u8');
  if (text.includes('#EXT-X-STREAM-INF')) {
    return { type: 'master', streams: parseMasterPlaylist(text, m3u8Url) };
  }
  return { type: 'segments', segments: parseSegmentPlaylist(text, m3u8Url) };
}

// ── Time-range segment selection ──────────────────────────────────────────
function selectSegments(segments, timeRange) {
  if (!timeRange) return { selected: segments, trimStart: 0, trimEnd: 0 };

  let cumulative = 0;
  let startIdx = -1, endIdx = -1;
  let trimStart = 0;

  for (let i = 0; i < segments.length; i++) {
    const segStart = cumulative;
    const segEnd = cumulative + segments[i].duration;

    if (startIdx === -1 && segEnd > timeRange.start) {
      startIdx = i;
      trimStart = timeRange.start - segStart;
    }
    if (segStart < timeRange.end) {
      endIdx = i;
    }
    cumulative += segments[i].duration;
  }

  if (startIdx === -1 || endIdx === -1) throw new Error('Time range outside video bounds');

  const selected = segments.slice(startIdx, endIdx + 1);
  const totalDuration = timeRange.end - timeRange.start;
  return { selected, trimStart, totalDuration };
}

// ── Parallel segment download ─────────────────────────────────────────────
async function downloadSegments(segments, tmpDir, concurrency = 8, onProgress) {
  const segFiles = [];
  let done = 0;
  const total = segments.length;

  for (let i = 0; i < total; i += concurrency) {
    const batch = segments.slice(i, i + concurrency);
    await Promise.all(batch.map((seg, bi) => {
      const idx = i + bi;
      const dest = path.join(tmpDir, `seg${String(idx).padStart(6, '0')}.ts`);
      segFiles[idx] = dest;
      return new Promise((resolve, reject) => {
        const proc = spawn('curl', [
          '-s', '-L', '--max-time', '120', '--retry', '3',
          ...CURL_HEADERS, seg.url, '-o', dest
        ]);
        proc.on('close', code => code === 0 ? resolve() : reject(new Error(`curl seg ${idx} exit ${code}`)));
        proc.on('error', reject);
      });
    }));
    done += batch.length;
    if (onProgress) onProgress(done, total);
  }
  return segFiles;
}

// ── ffmpeg remux ──────────────────────────────────────────────────────────
function remux(segFiles, outputFile, { trimStart = 0, totalDuration = 0 } = {}) {
  const tmpDir = path.dirname(segFiles[0]);
  const concatList = path.join(tmpDir, 'concat.txt');
  fs.writeFileSync(concatList, segFiles.map(f => `file '${f}'`).join('\n'));

  const ffmpeg = getFFmpegPath();
  const args = ['-f', 'concat', '-safe', '0', '-i', concatList];

  if (trimStart > 0) args.push('-ss', String(trimStart));
  if (totalDuration > 0) args.push('-t', String(totalDuration));

  args.push('-c', 'copy', '-movflags', '+faststart', '-y', outputFile);

  const res = spawnSync(ffmpeg, args, { stdio: 'pipe', timeout: 600000 });
  if (res.status !== 0) {
    const err = res.stderr?.toString().slice(-500) || '';
    throw new Error(`ffmpeg failed (exit ${res.status}): ${err}`);
  }
  return fs.statSync(outputFile).size;
}

// ── Full VOD download pipeline ────────────────────────────────────────────
async function downloadVOD({ m3u8Url, qualityLabel, outputFile, timeRange, onProgress, onStatus }) {
  onStatus?.('Fetching stream info...');
  const info = getQualities(m3u8Url);

  let segPlaylistUrl = m3u8Url;
  let qualities = [];

  if (info.type === 'master') {
    qualities = info.streams;
    // Pick requested quality or best
    const match = info.streams.find(s => s.url.includes(qualityLabel));
    segPlaylistUrl = match ? match.url : info.streams[0].url;
    onStatus?.(`Quality: ${match?.resolution || info.streams[0].resolution}`);
  }

  onStatus?.('Fetching segment list...');
  const segText = fetchM3U8(segPlaylistUrl);
  let segments = parseSegmentPlaylist(segText, segPlaylistUrl);
  const fullDuration = segments.reduce((s, seg) => s + seg.duration, 0);
  onStatus?.(`Full duration: ${formatDuration(fullDuration)} (${segments.length} segments)`);

  // Apply time range
  let trimOpts = { trimStart: 0, totalDuration: 0 };
  if (timeRange) {
    const result = selectSegments(segments, timeRange);
    segments = result.selected;
    trimOpts = { trimStart: result.trimStart, totalDuration: result.totalDuration };
    onStatus?.(`Time range: ${formatTimestamp(timeRange.start)} → ${formatTimestamp(timeRange.end)} (${segments.length} segments)`);
  }

  // Download segments
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'kick-'));
  try {
    onStatus?.('Downloading segments...');
    const segFiles = await downloadSegments(segments, tmpDir, 8, onProgress);

    onStatus?.('Remuxing to MP4...');
    const size = remux(segFiles, outputFile, trimOpts);
    return { size, duration: trimOpts.totalDuration || fullDuration, qualities };
  } finally {
    try { fs.rmSync(tmpDir, { recursive: true, force: true }); } catch (_) {}
  }
}

// ── Live stream recording ─────────────────────────────────────────────────
function recordLive(m3u8Url, outputFile, { duration, onStatus, onData }) {
  const ffmpeg = getFFmpegPath();
  const args = [
    '-user_agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
    '-headers', 'Referer: https://kick.com/\r\n',
    '-rw_timeout', '10000000',
    '-i', m3u8Url
  ];
  if (duration) args.push('-t', String(duration));
  args.push('-c', 'copy', '-movflags', '+faststart', '-y', outputFile);

  onStatus?.(`Recording live → ${outputFile}` + (duration ? ` (${formatDuration(duration)})` : ' (Ctrl+C to stop)'));

  return new Promise((resolve, reject) => {
    const proc = spawn(ffmpeg, args, { stdio: ['pipe', 'pipe', 'pipe'] });
    let lastLine = '';
    proc.stderr.on('data', chunk => {
      lastLine = chunk.toString().trim().split('\n').pop();
      const timeMatch = lastLine.match(/time=(\d+:\d+:\d+\.\d+)/);
      if (timeMatch) onData?.(timeMatch[1]);
    });
    proc.on('close', code => {
      if (code === 0 || code === 255) {
        const size = fs.existsSync(outputFile) ? fs.statSync(outputFile).size : 0;
        resolve({ size });
      } else reject(new Error(`ffmpeg exit ${code}`));
    });
    proc.on('error', reject);

    // Allow Ctrl+C to gracefully stop
    process.on('SIGINT', () => { proc.stdin.write('q'); });
  });
}

// ── Clip download ─────────────────────────────────────────────────────────
async function downloadClip(clipUrl, outputFile, onProgress) {
  // Clips are typically direct MP4 or short HLS
  if (clipUrl.endsWith('.mp4') || clipUrl.includes('.mp4?')) {
    // Direct MP4 download
    return new Promise((resolve, reject) => {
      const proc = spawn('curl', [
        '-s', '-L', '--max-time', '300', '--retry', '3',
        ...CURL_HEADERS, clipUrl, '-o', outputFile
      ]);
      proc.on('close', code => {
        if (code === 0) {
          resolve({ size: fs.statSync(outputFile).size });
        } else reject(new Error(`curl exit ${code}`));
      });
      proc.on('error', reject);
    });
  }
  // HLS clip
  return downloadVOD({ m3u8Url: clipUrl, qualityLabel: '720p60', outputFile, onProgress });
}

module.exports = {
  parseTimestamp, parseTimeRange, formatDuration, formatTimestamp,
  parseMasterPlaylist, parseSegmentPlaylist, getQualities, fetchM3U8,
  downloadVOD, recordLive, downloadClip, downloadSegments, remux,
};
